src/libcore/iter/traits/iterator.rs

```diff
@@ -1855,18 +1855,15 @@ pub trait Iterator {
     /// ```
     #[inline]
     #[stable(feature = "rust1", since = "1.0.0")]
-    fn all<F>(&mut self, f: F) -> bool where
+    fn all<F>(&mut self, mut f: F) -> bool where
```

Comment:
I'd like to keep these `Iterator::all` etc simplifications, there are examples of where this will be a big regression without them. We have already put in the work to make sure we thread internal iteration methods through various adaptors correctly, so to remove the use of `try_fold` here at the top level would be a step back.

This change only partly rolls back what we have worked on for internal iteration — other entry points to it still remain, like `sum`, `max`, `min`, `fold` and `try_fold`. And those remaining happen to be the most useful in stable Rust (because implementing `try_fold` is not possible in stable, so custom iterators can't fully take advantage of this `all` magic. It would also be inconsistent to partly remove it, I think we should keep it all.

Here's an example of what kind of improvement it can be (yes it's a microbenchmark) https://github.com/rust-itertools/itertools/pull/348  In this example it was `fold`, but it will also be `all` when `try_fold` is stable.

I think we should go through `fold`/`try_fold` everywhere we can, for the structural improvements that gives (bypasses complicated state machines that composed iterators might have in `.next()`)

Reply:
Of course the code inside Iterator::all is not beautiful to read as it is now, and if it could be simplified while keeping the same features, that would be fantastic.